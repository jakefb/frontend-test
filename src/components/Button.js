import React from 'react';
import PropTypes from 'prop-types';

/**
 * @public
 * Map of button types to class names
 */
export const buttonTypes = {
  submit: 'submit',
  button: 'button',
  reset: 'reset',
};

const star = (
  <svg className="c-icon" viewBox="0 0 15 15" role="presentation">
    <path d="M7.5 12.5l-5 2.5.833-5.417L0 5.833 5 5l2.5-5L10 5l5 .833-3.333 3.75L12.5 15z" />
  </svg>
);

const ButtonIcon = props => {
  if (props.isFavourite) {
    return (
      <React.Fragment>
        {star} {props.children}
      </React.Fragment>
    );
  }
  return props.children;
};

const Button = props => {
  let buttonClass;

  if (props.isFavourite) {
    buttonClass = 'c-button c-button-blue';
  } else {
    buttonClass = 'c-button';
  }

  if (props.isLink) {
    let rel;

    /** if isExternalLink is true prevent new page from having access to 'window.opener' */
    if (props.isExternalLink) {
      rel = 'external noopener noreferrer';
    }

    return (
      <a
        className={buttonClass}
        onClick={props.onClick}
        target={props.target}
        rel={rel}
        href={props.href}
      >
        <ButtonIcon isFavourite={props.isFavourite}>{props.children}</ButtonIcon>
      </a>
    );
  } else {
    return (
      <button className={buttonClass} onClick={props.onClick}>
        <ButtonIcon isFavourite={props.isFavourite}>{props.children}</ButtonIcon>
      </button>
    );
  }
};

ButtonIcon.propTypes = {
  children: PropTypes.node,

  /** isFavourite */
  isFavourite: PropTypes.bool,
};

Button.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,

  /** Determines if the button is disabled or not. */
  isDisabled: PropTypes.bool,

  /** Bind a function to fire when the button is clicked */
  onClick: PropTypes.func,

  /** Link href attribute */
  href: PropTypes.string,

  /** If true, sets appropriate `rel` values to prevent new page from having access to
   * `window.opener`. Should be used for links pointing at external sites. */
  isExternalLink: PropTypes.bool,

  /** Whether or not to render this button using an <a> tag */
  isLink: PropTypes.bool,

  /** The HTML tabIndex attribute value */
  tabIndex: PropTypes.number,

  /** The `target` attribute to use on the anchor element (ignored unless `isLink` is `true`) */
  target: PropTypes.string,

  /** The `title` attribute */
  title: PropTypes.string,

  /** isFavourite */
  isFavourite: PropTypes.bool,
};

Button.defaultProps = {
  isDisabled: false,
  isExternalLink: false,
  isLink: false,
  tabIndex: 0,
  type: buttonTypes.button,
};

export default Button;
