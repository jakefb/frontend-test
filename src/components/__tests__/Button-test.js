/**
 * These tests are pre-configured with Jest (https://jestjs.io/docs/en/getting-started),
 * and use sinon.js (https://sinonjs.org) for mocking, and React Testing Library
 * (https://testing-library.com/docs/react-testing-library/intro) for testing React
 * components. You are welcome to use these tools, replace with your preferred tools,
 * or write pseudo code to communicate your ideas.
 */
import 'react-testing-library/cleanup-after-each';
import React from 'react';
import Button from '../Button';
import sinon from 'sinon';
import { render, fireEvent } from 'react-testing-library';

describe('<Button/>', () => {
  it('should render as a button element when no type is provided', () => {
    const { getByText } = render(<Button>Test button</Button>);
    var buttonElement = getByText('Test button');
    expect(buttonElement.nodeName).toBe('BUTTON');
    expect(buttonElement).toMatchSnapshot();
  });

  it('should render as a link element when isLink flag is set', () => {
    const { getByText } = render(<Button isLink={true}>Test link</Button>);
    var buttonElement = getByText('Test link');
    expect(buttonElement.nodeName).toBe('A');
    expect(buttonElement).toMatchSnapshot();
  });

  it('should set the `target` prop as the `target` attribute if rendering a link', () => {
    const { getByText } = render(
      <Button isLink={true} href="https://www.xero.com/nz/" target="_blank">
        Test link with target
      </Button>,
    );
    var buttonElement = getByText('Test link with target');
    expect(buttonElement.getAttribute('target')).toBe('_blank');
    expect(buttonElement).toMatchSnapshot();
  });

  it('should render as a link element with `rel="external noopener noreferrer"` if the `isExternalLink` prop is true', () => {
    const { getByText } = render(
      <Button isLink={true} href="https://www.xero.com/nz/" isExternalLink={true}>
        Test external link
      </Button>,
    );
    var buttonElement = getByText('Test external link');
    expect(buttonElement.getAttribute('rel')).toBe('external noopener noreferrer');
    expect(buttonElement).toMatchSnapshot();
  });

  it('should render as a link element with existing `rel` value intact when `isExternalLink` prop is false', () => {
    const { getByText } = render(
      <Button isLink={true} href="https://www.xero.com/nz/" isExternalLink={false} rel="nofollow">
        Test internal link
      </Button>,
    );
    var buttonElement = getByText('Test internal link');
    expect(buttonElement.getAttribute('rel')).toBe(null);
    expect(buttonElement).toMatchSnapshot();
  });

  it('should be able to receive and handle an onClick callback', () => {
    var testMethod = {
      method: function(test) {
        return test;
      },
    };
    var mock = sinon.mock(testMethod);
    mock
      .expects('method')
      .once()
      .withExactArgs('foo bar');
    const { getByText } = render(
      <Button
        onClick={() => {
          testMethod.method('foo bar');
        }}
      >
        Test onClick
      </Button>,
    );
    var buttonElement = getByText('Test onClick');
    fireEvent.click(buttonElement);
    mock.verify();
    expect(buttonElement).toMatchSnapshot();
  });

  it('should support an href attribute if the button is a link and no click handler has been defined', () => {
    const { getByText } = render(
      <Button isLink={true} href="https://www.xero.com/nz/">
        Link with href attribute
      </Button>,
    );
    var buttonElement = getByText('Link with href attribute');
    expect(buttonElement.getAttribute('href')).toBe('https://www.xero.com/nz/');
    expect(buttonElement).toMatchSnapshot();
  });

  it('should be able to handle click events for links', () => {
    const { getByText } = render(
      <Button isLink={true} href="https://www.xero.com/nz/">
        Link with href attribute
      </Button>,
    );
    var buttonElement = getByText('Link with href attribute');
    var eventTargetHref = null;
    buttonElement.addEventListener('click', event => {
      eventTargetHref = event.target.href;
    });
    fireEvent.click(buttonElement);
    expect(eventTargetHref).toBe('https://www.xero.com/nz/');
    expect(buttonElement).toMatchSnapshot();
  });
});
