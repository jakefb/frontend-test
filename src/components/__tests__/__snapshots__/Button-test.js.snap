// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should be able to handle click events for links 1`] = `
<a
  class="c-button"
  href="https://www.xero.com/nz/"
>
  Link with href attribute
</a>
`;

exports[`<Button/> should be able to receive and handle an onClick callback 1`] = `
<button
  class="c-button"
>
  Test onClick
</button>
`;

exports[`<Button/> should render as a button element when no type is provided 1`] = `
<button
  class="c-button"
>
  Test button
</button>
`;

exports[`<Button/> should render as a link element when isLink flag is set 1`] = `
<a
  class="c-button"
>
  Test link
</a>
`;

exports[`<Button/> should render as a link element with \`rel="external noopener noreferrer"\` if the \`isExternalLink\` prop is true 1`] = `
<a
  class="c-button"
  href="https://www.xero.com/nz/"
  rel="external noopener noreferrer"
>
  Test external link
</a>
`;

exports[`<Button/> should render as a link element with existing \`rel\` value intact when \`isExternalLink\` prop is false 1`] = `
<a
  class="c-button"
  href="https://www.xero.com/nz/"
>
  Test internal link
</a>
`;

exports[`<Button/> should set the \`target\` prop as the \`target\` attribute if rendering a link 1`] = `
<a
  class="c-button"
  href="https://www.xero.com/nz/"
  target="_blank"
>
  Test link with target
</a>
`;

exports[`<Button/> should support an href attribute if the button is a link and no click handler has been defined 1`] = `
<a
  class="c-button"
  href="https://www.xero.com/nz/"
>
  Link with href attribute
</a>
`;
